File: 10. Regular Expression Matching.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        bool[,] dp = new bool[s.Length + 1, p.Length + 1];
        dp[s.Length, p.Length] = true;

        for (int i = s.Length; i >= 0; i--){
            for (int j = p.Length - 1; j >= 0; j--){
                bool firstMatch = (i < s.Length && (p[j] == s[i] || p[j] == '.'));
                
                if (j + 1 < p.Length && p[j + 1] == '*') {
                    dp[i, j] = dp[i, j + 2] || firstMatch && dp[i + 1, j];
                } else {
                    dp[i, j] = firstMatch && dp[i + 1, j + 1];
                }
            }
        }
        
        return dp[0, 0];
    }
}

File: 10. Regular Expression Matching.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();
        vector<vector<bool>> dp(m + 1, vector<bool>(n + 1, false));
        dp[m][n] = true; // Both strings are empty
        
        for (int i = m; i >= 0; i--){
            for (int j = n - 1; j >= 0; j--){
                bool firstMatch = (i < m && (p[j] == s[i] || p[j] == '.'));
                
                if (j + 1 < n && p[j + 1] == '*') {
                    dp[i][j] = dp[i][j + 2] || firstMatch && dp[i + 1][j];
                } else {
                    dp[i][j] = firstMatch && dp[i + 1][j + 1];
                }
            }
        }
        
        return dp[0][0];
    }
};

